ops/model.py

```diff
@@ -1326,6 +1327,145 @@ def list_files(self, path: str, *, pattern: str = None,
         """
         return self._pebble.list_files(path, pattern=pattern, itself=itself)
 
+    def push_path(self,
+                  source_path: typing.Union[typing.Union[str,
+                                                         Path],
+                                            typing.List[typing.Union[str,
+                                                                     Path]]],
+                  dest_dir: typing.Union[str,
+                                         Path]):
+        """Push all the things.
+
+        TODO: write this docstring
+        """
+        source_paths = source_path
+        if not isinstance(source_paths, list):
+            source_paths = [source_paths]
+        source_paths = [str(p) for p in source_paths]
+        dest_dir = str(dest_dir)
+
+        def local_list(source_path: str):
+            paths = [source_path]
+            if os.path.isdir(source_path):
+                paths = os.listdir(source_path)
+            files = []
+            for name in paths:
+                fpath = os.path.join(source_path, name)
+                info = os.lstat(fpath)
+                if stat.S_ISDIR(info.st_mode):
+                    ftype = pebble.FileType.DIRECTORY
+                elif stat.S_ISREG(info.st_mode):
+                    ftype = pebble.FileType.FILE
+                elif stat.S_ISLINK(info.st_mode):
+                    ftype = pebble.FileType.SYMLINK
+                else:
+                    continue
+
+                if os.name == 'nt':
+                    files.append(
+                        pebble.FileInfo(
```

Comment:
This (and the other) would be great to have in some helper function. `build_fileinfo` or whatever, which uses kwargs.

This is awfully dependent on `ops.pebble.FileInfo` never changing the order of the args, and putting it away behind a factory helper method and/or naming the args would anyone reading the code instead of being confronted with a list of 10 variables in a particular order and maybe wondering what's so important about it, rather than just seeing a call to `_build_fileinfo` or whatever.

Incidentally, if `pathlib` were used here, you could type assert `WindowsPath` and built it there, which helps readability versus knowing that `os.name == 'nt'` means Windows.

Reply:
I just looked at pebble code - and it doesn't do anything special to catch/handle non-unix filepaths.  So if we want to support windows clients pushing, then we are going to have to convert the filepaths back and forth (windows <--> unix).  Or we could just punt for now and say these pebble-based api funcs only work with unix-based charms.

Reply:
Up to you, really. I meant more in the sense that the UID/GID from fat32 filesystems on Linux are ultimately not meaningful.

As mentioned, kind of an edge case anyway. Only asked because there's already the case for "if it's Windows..."

Either way, `pebble.FileInfo` construction would be great in a factory

Reply:
I'd ignore the Windows  compatibility for now, the Juju team are shortly going to deprecate support for running charms on Windows (Juju 3.0+), so I suspect the use-case will dissolve.

Reply:
:+1: this makes life much easier in other places